File: 45jumpGameII.py
# 45. Jump Game II

# https://www.youtube.com/watch?v=dJ7sWiOoK7g

from collections import deque

class Solution:
    def jump(self, nums: List[int]) -> int:
        
        res = 0   
        l = r = 0
                
        while r<len(nums)-1:
            
            farthest = 0
            
            for i in range(l, r+1):
                farthest = max(farthest, i+nums[i])
                
            l = r+1
            r = farthest
            res+=1
        
        return res
                
                
#         MY SOLUTION THAT IS CORRECT BUT EXCEEDS TIME LIMIT                
        
#         n = 0
#         group = deque([0])
#         prev = []
        
#         while(len(nums)-1 not in group):
            
#             for i in range(len(group)):
                
#                 ind = group.popleft()
#                 prev.append(ind)
#                 step = nums[ind]
                
                
#                 while step>0:
#                     if ind+step<len(nums) and ind+step not in group and ind+step not in prev:
#                         group.append(ind+step)
#                     step-=1
                       
#             n+=1
            
#         return n

File: 45_jumpGameII.cpp
// 45. Jump Game II

class Solution {
public:
    // DYNAMIC PROGRAMMING
    int jump(vector<int>& nums) {
        
        vector<int> res;
        
        for(int i=0;i<nums.size();i++)
            res.push_back(INT_MAX-1);
        
        res[nums.size()-1] = 0;
        
        for(int i=nums.size()-1;i>-1;i--){
            for(int j=i+1;j<i+nums[i]+1;j++){
                if(j<nums.size())
                    res[i] = min(res[i], 1+res[j]);  
            }
        }
        
        return res[0];
    }
};
